etcd prefix fix and migration for 4.3

awly opened the pull request:
See https://github.com/gravitational/teleport/issues/2883 for context.
This PR makes the etcd backend _actually_ use the prefix specified in config file. It also migrates the data from the old prefix as needed.

Specific changes are:
1. always use the `teleport.backend.prefix` field for etcd storage (instead of hardcoded `/teleport`)
2. at startup, overwrite any config prefix data with data from `/teleport` iff (a) `/teleport` has some data and (b) config prefix is empty (initial 4.2->4.3 upgrade) or has older data than `/teleport` (upgrade 4.2->4.3, downgrade 4.3->4.2, make some data changes and upgrade 4.2->4.3 again)

This startup copying is blocking - teleport won't start until it's complete.

All of this assumes that there's only 1 auth server during an upgrade, per our upgrade docs.

Commit:
etcd: propagate the prefix from config

Prefix-handling code was using a hardcoded prefix (`/teleport`) instead
of the prefix specified in config. Use the correct config prefix and add
a test.

Commit:
etcd: double-write to legacy /teleport prefix

This double-writing will only exist in 4.3 and the config prefix is
still source-of-truth for reads.
Double-writes are necessary to allow users to downgrade to 4.2 and back
up again without losing data.

Commit:
etcd: copy data from the legacy prefix to config prefix at startup

This code is only intended for 4.3 and will be removed in 4.4.
The goal is to automatically move data to the correct etcd prefix that
the customer has specified in their config. Prior to 4.3, their data was
always written to `/teleport` prefix, ignoring the config.

The copy happens iff:
- /teleport is not empty
- config prefix is empty OR older than /teleport

Commit:
etcd: fix etcd double-writing and add tests

Using the wrong prefix in some double-write methods, oops.
Added tests to make sure it doesn't happen again.

klizhentas commented:
@awly a couple of questions:

* Are double writes really necessary  - assuming migration has happened on the first start?
* With double write strategy, would not it mean that the second write to legacy index will generate a newer revision number leading to second copy?

I wonder if as a strategy, we can only migrate once under lock as usual here:

https://github.com/gravitational/teleport/blob/master/lib/auth/init.go#L438

And then remove all the legacy code from the backend.

awly commented:
> Are double writes really necessary - assuming migration has happened on the first start?

I did this in case a user finds some catastrophic bug after an upgrade and needs to revert back to 4.2.
Double-writing preserves any changes made while they ran 4.3.

> With double write strategy, would not it mean that the second write to legacy index will generate a newer revision number leading to second copy?

Hmm, yes, the legacy prefix will always be 1 revision ahead. I think I should swap the order. Good catch!

> I wonder if as a strategy, we can only migrate once under lock as usual here:

Oh, I didn't see that before.
This migration is slightly different, in that it's specific to a backend type and needs raw access to etcd and not our backend wrapper (because the wrapper hides prefixes entirely). I'm not sure we want to expose raw etcd connection and logic up to `lib/auth`.

klizhentas commented:
> I did this in case a user finds some catastrophic bug after an upgrade and needs to revert back to 4.2.
Double-writing preserves any changes made while they ran 4.3.

Instead of pursuing this strategy (and my main concern is complexity of double writes), can we instead ask folks to use backup/restore of resources to roll back the new data? 

> Oh, I didn't see that before.
This migration is slightly different, in that it's specific to a backend type and needs raw access to etcd and not our backend wrapper (because the wrapper hides prefixes entirely). I'm not sure we want to expose raw etcd connection and logic up to lib/auth.

One way of doing that is calling backend's `Migrate` method. `auth/init` has backend already initialized, and the method can be `Nop` for most of the backends most of the time.

Commit:
Add a Migrate method to backend.Backend

Unify migrations and expose them to the calling code at startup.
All backends except for etcd implement a nop migration.

awly commented:
> Instead of pursuing this strategy (and my main concern is complexity of double writes), can we instead ask folks to use backup/restore of resources to roll back the new data?

I'm all for reducing complexity. To clarify, your suggestion is to add a note in release notes under "upgrading" section saying something like: "If you use etcd backend and decide to downgrade from 4.3 back to 4.2, do a manual backup/restore to keep recent changes"?
If that's the case, I'll remove all double-writing.

> One way of doing that is calling backend's Migrate method. auth/init has backend already initialized, and the method can be Nop for most of the backends most of the time.

Added a `Migrate` method that gets called on auth server startup when it initializes the backend.

klizhentas commented:
> I'm all for reducing complexity. To clarify, your suggestion is to add a note in release notes under "upgrading" section saying something like: "If you use etcd backend and decide to downgrade from 4.3 back to 4.2, do a manual backup/restore to keep recent changes"?
If that's the case, I'll remove all double-writing.

Exactly, let's do just that.

> Added a Migrate method that gets called on auth server startup when it initializes the backend.

Perfect, thanks!

Commit:
Revert etcd double-writing

Per discussion in https://github.com/gravitational/teleport/pull/3798
this is unnecessary complexity.

awly commented:
@klizhentas done, PTAL

awly commented:
retest this please

klizhentas commented:
@awly thanks looks good to me now

awly commented:
retest this please

russjones commented:
@gravitational-jenkins retest this please

Commit:
etcd: create a backend per-test instead of per-suite

This avoids data races between watchers and test setup.
